### PR TITLE
Reorganize vmc

### PR DESCRIPTION
**Pull request by NIC619**
### How was it fixed?
Clean up VMC handler, log handler. And move ShardTracker away from VMC handler.
ShardTracker and log handler will be used to record and fetch logs(CollationHeader info) while VMC handler is used to interact with VMC contract(get contract info, deposit, withdraw, etc).

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

File: evm/vm/forks/sharding/vmc_handler.py [hunk 1]
```diff
@@ -1,271 +1,179 @@
...
-                break
-            self.unchecked_logs.append(log_entry)
-        self.current_score = log_entry['score']
-        return log_entry
-
-
-class VMC(Contract):
-
-    logger = logging.getLogger("evm.chain.sharding.VMC")
+def mk_call_context(sender_address,
```
Comment (hwwhww) [inline evm/vm/forks/sharding/vmc_handler.py]: Renaming suggestion: all `mk` -> `make` to fit the py-evm convention.

File: evm/vm/forks/sharding/vmc_handler.py [hunk 2]
```diff
@@ -1,271 +1,179 @@
...
-    def get_default_sender_address(self):
-        return self.default_sender_address
-
-    def set_shard_tracker(self, shard_id, shard_tracker):
-        self.shard_trackers[shard_id] = shard_tracker
+    #
+    # property
+    #
+    @property
+    def get_privkey(self):
```
Comment (hwwhww) [inline evm/vm/forks/sharding/vmc_handler.py]: Just name it `privkey`.

File: evm/vm/forks/sharding/vmc_handler.py [hunk 3]
```diff
@@ -1,271 +1,179 @@
...
+        return self._privkey
 
-    def get_shard_tracker(self, shard_id):
-        if shard_id not in self.shard_trackers:
-            raise UnknownShard('Shard {} is not tracked'.format(shard_id))
-        return self.shard_trackers[shard_id]
+    @property
+    def get_sender_address(self):
```
Comment (hwwhww) [inline evm/vm/forks/sharding/vmc_handler.py]: ditto

File: evm/vm/forks/sharding/vmc_handler.py [hunk 4]
```diff
@@ -1,271 +1,179 @@
...
+        return self._sender_address
 
-    # TODO: currently just calls `shard_tracker.get_next_log`
-    def get_next_log(self, shard_id):
-        shard_tracker = self.get_shard_tracker(shard_id)
-        return shard_tracker.get_next_log()
+    @property
+    def get_config(self):
```
Comment (hwwhww) [inline evm/vm/forks/sharding/vmc_handler.py]: ditto

File: evm/vm/forks/sharding/vmc_handler.py [hunk 5]
```diff
@@ -1,271 +1,179 @@
...
-            self.unchecked_logs.append(log_entry)
-        self.current_score = log_entry['score']
-        return log_entry
-
-
-class VMC(Contract):
-
-    logger = logging.getLogger("evm.chain.sharding.VMC")
+def mk_call_context(sender_address,
+                    gas=None,
```
Comment (hwwhww) [inline evm/vm/forks/sharding/vmc_handler.py]: `gas` is required now.

File: evm/vm/forks/sharding/log_handler.py [hunk 6]
```diff
@@ -1,25 +1,23 @@
...
 def get_recent_block_hashes(w3, history_size):
     block = w3.eth.getBlock('latest')
-    recent_hashes = [block['hash']]
     # initialize the list of recent hashes
+    recent_hashes = [block['hash']]
+
     for _ in range(history_size - 1):
         # break the loop if we hit the genesis block.
         if block['number'] == 0:
+            recent_hashes.append(block['hash'])
```
[the remark this answers is not in the document]
Comment (NIC619) [reply evm/vm/forks/sharding/log_handler.py]: You are right. I will fix it.

File: evm/vm/forks/sharding/log_handler.py [hunk 7]
```diff
@@ -110,10 +79,11 @@ def get_new_logs(self, address=None, topics=None):
         from_block_number = self.w3.eth.getBlock(from_block_hash)['number']
         to_block_number = self.w3.eth.getBlock(to_block_hash)['number']
 
-        filter_params = self.mk_filter_params(
```
[the remark this answers is not in the document]
Comment (NIC619) [reply evm/vm/forks/sharding/log_handler.py]: `getLogs` accepts `None` as passed in value of parameters so I figured filter arguments can be passed in directly.
```python
return self.w3.eth.getLogs(
    {
        'fromBlock': from_block_number,
        'toBlock': to_block_number,
        'address': address,
        'topics': topics,
     }
)
```

**Comment by NIC619** [comment]
@hwwhww @mhchia 
Updated with PR feedbacks.
Please have another look. Thanks.

**Pull request merged**